uast/diff/diff.go

```diff
@@ -0,0 +1,287 @@
+type sameDecision struct{ basicDecision }
+type replaceDecision struct{ basicDecision }
+type matchDecision struct{ basicDecision }
+type permuteDecision struct {
+	basicDecision
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func min(self, candidate decisionType) decisionType {
+	if self.cost() > candidate.cost() {
+		return candidate
+	} else {
+		return self
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decisionType
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decisionType),
+		make(map[ID]int),
+		nil,
+	}
+}
+
+// caches (for perf) tree size (node count)
+func (ds *cacheStorage) nodeSize(n nodes.Node) int {
+	label := nodes.UniqueKey(n)
+	if cnt, ok := ds.sizes[label]; ok {
+		return cnt
+	}
+	ret := nodes.Count(n, nodes.KindsNotNil)
+	ds.sizes[label] = ret
+	return ret
+}
+
+// find the cheapest way to naively match src and dst and return the action with its combined cost
+func (ds *cacheStorage) decideAction(src, dst nodes.Node) decisionType {
+	label := keyType{nodes.UniqueKey(src), nodes.UniqueKey(dst)}
+
+	if val, ok := ds.decisions[label]; ok {
+		return val
+	}
+
+	// one can always just create the dst ignoring the src
+	cost := ds.nodeSize(dst) + 1
+
+	var bestDecision decisionType
+	bestDecision = replaceDecision{basicDecision{cost}}
+
+	if nodes.KindOf(src) != nodes.KindOf(dst) {
+		ds.decisions[label] = bestDecision
+		return bestDecision
+	}
+
+	switch src.(type) {
+
+	case nodes.Value:
+		src, dst := src.(nodes.Value), dst.(nodes.Value)
+		cost = 0
+		if src == dst {
+			bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+		} else {
+			cost = 1
+			bestDecision = min(bestDecision, replaceDecision{basicDecision{cost}})
+		}
+
+	case nodes.Object:
+		src, dst := src.(nodes.Object), dst.(nodes.Object)
+		cost = 0
+
+		keys := make(map[string]bool)
+		iterate := func(keyset nodes.Object) {
+			for key := range keyset {
+				if in := keys[key]; !in {
+					keys[key] = true
+					cost += ds.decideAction(src[key], dst[key]).cost()
+				}
+			}
+		}
+		iterate(src)
+		iterate(dst)
+
+		if cost == 0 {
+			bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+		} else {
+			bestDecision = min(bestDecision, matchDecision{basicDecision{cost}})
+		}
+
+	case nodes.Array:
+		src, dst := src.(nodes.Array), dst.(nodes.Array)
+		cost = 0
+		if len(src) == len(dst) && func() int {
+			sum := 0
+			for i := range src {
+				sum += ds.decideAction(src[i], dst[i]).cost()
+			}
+			return sum
+		}() == 0 {
+			bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+			break
+		}
+
+		if len(src) != len(dst) {
+			cost = 2
+		}
+
+		if len(src) < len(dst) {
+			arr := make(nodes.Array, len(dst))
+			copy(arr, src)
+			src = arr
+		} else {
+			arr := make(nodes.Array, len(src))
```

Comment:
This branch will trigger for the case of `len(src) == len(dst)` and different children (`cost != 0`).

Reply:
right. fixed.